### PR TITLE
feat(sidebar): added sidebar component

### PR DESCRIPTION
**Pull request by mcoker**
fixes https://github.com/patternfly/patternfly/issues/3653

This adds a basic sidebar component. In a follow up sprint, I would like to add responsive support to the stack layout, split layout, gutters, sticky, and right panel modifiers.

### CODE REVIEW

**Review by mcarrano — COMMENTED** [review]
Looks good @mcoker , but I have a few questions.

* What's the purpose of the Stacked example? Seems counter-intuitive to have a stacked siderbar.

* It would be great if you can either add some documentation or something visual that differentiates your examples better. It's hard for me to tell what's different about "basic," "gutter," and "split" for example.

* Why do we have 2 sticky panel examples? Are they intended to be different?

**Review by DaoDaoNoCode — COMMENTED** [review]
@mcoker Looks great! And I agree with @mcarrano . I cannot tell the difference between basic and split sidebar, and I can see there is a space between panel and content for gutter sidebar, but it's not obvious. Is there a way to make it more obvious? Like adding background/border/divider to the example?

File: src/patternfly/components/Sidebar/examples/Sidebar.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+{{#> sidebar}}
+  {{#> sidebar-panel}}
+    Sidebar panel
+  {{/sidebar-panel}}
+  {{#> sidebar-content}}
+    {{#> content}}
+      <p>Moves the panel to the right by placing the panel after the content in the HTML.</p>
+      <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse dapibus nulla id augue dictum commodo. Donec mollis arcu massa, sollicitudin venenatis est rutrum vitae. Integer pulvinar ligula at augue mollis, ac pulvinar arcu semper. Maecenas nisi lorem, malesuada ac lectus nec, porta pretium neque. Ut convallis libero sit amet metus mattis, vel facilisis lorem malesuada. Duis consectetur ante sit amet magna efficitur, a interdum leo vulputate.</p>
+      <p>Praesent at odio nec sapien ultrices tincidunt in non mauris. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Duis consectetur nisl quis facilisis faucibus. Sed eu bibendum risus. Suspendisse porta euismod tortor, at elementum odio suscipit sed. Cras eget ultrices urna, ac feugiat lectus. Integer a pharetra velit, in imperdiet mi. Phasellus vel hendrerit velit. Vestibulum ut augue vitae erat vulputate bibendum a ut magna.</p>
+    {{/content}}
```
Comment (DaoDaoNoCode) [inline src/patternfly/components/Sidebar/examples/Sidebar.md:86]: Do we need to place the content before panel in this example?

File: src/patternfly/components/Sidebar/examples/Sidebar.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  {{/sidebar-content}}
+{{/sidebar}}
+```
+
+### Stack
+```hbs
+{{#> sidebar sidebar--modifier="pf-m-stack"}}
+  {{#> sidebar-panel}}
+    Sidebar panel
+  {{/sidebar-panel}}
```
Comment (DaoDaoNoCode) [inline src/patternfly/components/Sidebar/examples/Sidebar.md:48]: The panel here seems not aligned to the top of `__main`, is it designed to do so?

File: src/patternfly/components/Sidebar/examples/Sidebar.md (hunk 1 above)
Comment (mcoker) [reply src/patternfly/components/Sidebar/examples/Sidebar.md:83]: Yep! Updated.

File: src/patternfly/components/Sidebar/examples/Sidebar.md (hunk 2 above)
Comment (mcoker) [reply src/patternfly/components/Sidebar/examples/Sidebar.md:48]: The way sticky works, it's not really designed to do that, it's designed to stick to the top of the closest ancestor with `overflow` set to scroll. I moved the local styles that define height and overflow to another workspace example element though, hopefully that behaves more like we would expect.

**Comment by mcoker** [comment]
>  I can see there is a space between panel and content for gutter sidebar, but it's not obvious. Is there a way to make it more obvious? Like adding background/border/divider to the example?

Updated the example so there is more content in the panel and the gutter is more visible.

**Comment by mcoker** [comment]
> What's the purpose of the Stacked example?

It's basically the responsive version. By default the layout is stacked up to the medium breakpoint, then it changes to a side by side layout. This allows the user to specify the layout is stacked by default, then split at the XL breakpoint. Or split by default, stacked at medium, split at XL, etc.

> It would be great if you can either add some documentation or something visual that differentiates your examples better.

Added descriptions in the example.

> Why do we have 2 sticky panel examples? Are they intended to be different?

Nope! Copy 'n paste error.

**Comment by mcarrano** [comment]
Thanks for the clarifications @mcoker

File: src/patternfly/components/Sidebar/examples/Sidebar.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+## Documentation
+
+### Usage
+| Class | Applied to | Outcome |
+| -- | -- | -- |
+| `.pf-c-sidebar` | `<div>` | Initiates the sidebar component. **Required** |
+| `.pf-c-simple-list__main` | `<div>` | Initiates the sidebar main element. **Required** |
+| `.pf-c-simple-list__panel` | `<div>` | Initiates the sidebar panel element. **Required** |
+| `.pf-c-simple-list__content` | `<div>` | Initiates the sidebar content element. **Required** |
```
Comment (mattnolting) [inline src/patternfly/components/Sidebar/examples/Sidebar.md:163]:
```suggestion
| `.pf-c-sidebar__main` | `<div>` | Initiates the sidebar main element. **Required** |
| `.pf-c-sidebar__panel` | `<div>` | Initiates the sidebar panel element. **Required** |
| `.pf-c-sidebar__content` | `<div>` | Initiates the sidebar content element. **Required** |
```

File: src/patternfly/components/Sidebar/sidebar.scss [hunk 4]
```diff
@@ -0,0 +1,184 @@
+$pf-c-sidebar--breakpoint-map--width: build-breakpoint-map("base", "sm", "md", "lg", "xl", "2xl");
+$pf-c-sidebar__panel--list--width: ("default", 25, 33, 50, 66, 75, 100);
+
+
+.pf-c-sidebar {
+  --pf-c-sidebar--BackgroundColor: var(--pf-global--BackgroundColor--100);
+
+  // Main
+  --pf-c-sidebar__main--FlexDirection: column;
+  --pf-c-sidebar__main--md--FlexDirection: row;
```
Comment (mattnolting) [inline src/patternfly/components/Sidebar/sidebar.scss:10]: 👍

File: src/patternfly/components/Sidebar/sidebar-content.hbs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<div class="pf-c-sidebar__content{{#if sidebar-content--modifier}} {{sidebar-content--modifier}}{{/if}}"
+  {{#if sidebar-content--attribute}}
+    {{{sidebar-content--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
Comment (mattnolting) [inline src/patternfly/components/Sidebar/sidebar-content.hbs:5]:
```suggestion
  {{> @partial-block}}
```

File: src/patternfly/components/Sidebar/sidebar-main.hbs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<div class="pf-c-sidebar__main{{#if sidebar-main--modifier}} {{sidebar-main--modifier}}{{/if}}"
+  {{#if sidebar-main--attribute}}
+    {{{sidebar-main--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
Comment (mattnolting) [inline src/patternfly/components/Sidebar/sidebar-main.hbs:5]:
```suggestion
  {{> @partial-block}}
```

File: src/patternfly/components/Sidebar/sidebar-panel.hbs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<div class="pf-c-sidebar__panel{{#if sidebar-panel--modifier}} {{sidebar-panel--modifier}}{{/if}}"
+  {{#if sidebar-panel--attribute}}
+    {{{sidebar-panel--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
Comment (mattnolting) [inline src/patternfly/components/Sidebar/sidebar-panel.hbs:5]:
```suggestion
  {{> @partial-block}}
```

**Review by mattnolting — CHANGES_REQUESTED** [review]
This looks great! Super simple and flexible. Just a few housekeeping notes.

**Comment by mcoker** [comment]
@DaoDaoNoCode @mattnolting updated!

**Review by mattnolting — APPROVED** [review]
LPTM 🎉

**Review by DaoDaoNoCode — APPROVED** [review]
Looks great to me, Thank you @mcoker !

**Review by mcarrano — APPROVED** [review]
Looks good. Please go ahead and merge.

**Pull request merged**